File: ScanLineRender/ScanLineRender/Primitive.c
/*
//  Primitive.c
//  ScanLineRender
//
//  Created by Thomas Dickerson on 1/25/16.
//  Copyright © 2016 StickFigure Graphic Productions. All rights reserved.
*/

#include "Primitive.h"
#include <math.h>
#include <string.h>
#include "rb_tree/misc.h"

void makeLine(Primitive *o, Color c, const Edge e){
	Primitive tmp;
	Point **const b = SafeMalloc(sizeof(Edge));
	memcpy(b, e, sizeof(Edge));
	INIT_PRIM(tmp, c, 1, b);
	*o = tmp;
}
void makeTri(Primitive *o, Color c, const Edge e1, const Edge e2, const Edge e3){
	Primitive tmp;
	Point **const b = SafeMalloc(4*sizeof(Point*)),
	**bn = b;
	assert(e1[END] == e2[START]
		   && e2[END] == e3[START]
		   && e3[END] == e1[START]
		   && "Not a closed line-loop");
	memcpy(bn++, e1, sizeof(Point*));
	memcpy(bn++, e2, sizeof(Point*));
	memcpy(bn, e3, sizeof(Edge));
	INIT_PRIM(tmp, c, 3, b);
	*o = tmp;
}

void makeQuad(Primitive *o, Color c, const Edge e1, const Edge e2, const Edge e3, const Edge e4){
	Primitive tmp;
	Point **const b = SafeMalloc(5*sizeof(Point*)),
	**bn = b;
	assert(e1[END] == e2[START]
		   && e2[END] == e3[START]
		   && e3[END] == e4[START]
		   && e4[END] == e1[START]
		   && "Not a closed line-loop");
	memcpy(bn++, e1, sizeof(Point*));
	memcpy(bn++, e2, sizeof(Point*));
	memcpy(bn++, e3, sizeof(Point*));
	memcpy(bn, e4, sizeof(Edge));
	INIT_PRIM(tmp, c, 4, b);
	*o = tmp;
}

float getZForXY(const Primitive *p, float x, float y){
	Point **const boundary = p->boundary;
	if(p->arity == 1){
		const Point *const vs = boundary[START],
		* ve = boundary[END];
		
		const float sx = vs->x,
		sy = vs->y,
		sz = vs->z,
		
		dx = ve->x - sx,
		dy = ve->y - sy,
		dz = ve->z - sz;
		
		/* Theoretically these should be the same, but
		 // we'll average for robustness */
		const float xNumer = dz * (x - sx),
		yNumer = dz * (y - sy);
		const float xEst = (dx == 0) ? min(0,  dz) : (xNumer / dx),
		yEst = (dy == 0) ? min(0,  dz) : (yNumer / dy);
		
		return sz + (xEst + yEst) / 2;
	} else {
		Point **const e1 = boundary,
		**const e2 = e1 + 1;
		
		const Point * us = e1[START],
		* ue = e1[END],
		* vs = e2[START],
		* ve = e2[END];
		
		const float ux = us->x - ue->x,
		uy = us->y - ue->y,
		uz = us->z - ue->z,
		vx = vs->x - ve->x,
		vy = vs->y - ve->y,
		vz = vs->z - ve->z;
		
		const float nx = uy * vz - uz * vy,
		ny = uz * vx - ux * vz,
		nz = ux * vy - uy * vx;
		
		const float d = -nx * us->x
		-ny * us->y
		-nz * us->z,
		numer = (-d - nx * x - ny * y);
		
		return (nz == 0) ? ((numer > 0) ? HUGE_VAL : -HUGE_VAL) : (numer / nz) ;
	}
	
}


void transformPrimitive(const Transformation * txForm, const Primitive *p, Primitive *o){
	const TransformationF f = txForm->f;
	const void * state = txForm->state;
	Point **const pBoundary = p->boundary;
	Point **const oBoundary = o->boundary;
	size_t i, iMax = p->arity;
	for(i = 0; i <= iMax; ++i){
		f(pBoundary[i],oBoundary[i],state);
	}
}


int32_t hashPrim(const Primitive *p){
	size_t i, iMax = p->arity; int32_t ret = 0;
	for(i = 0; i <= iMax; ++i){
		ret ^= HASH_POINT(*(p->boundary[i]));
	}
	ret ^= (p->color << 8) | 0xFF;
	return ret;
}

File: src/net/cemetech/sfgp/scanline/geom/Primitive.java
package net.cemetech.sfgp.scanline.geom;

import java.awt.Color;
import java.util.LinkedList;

import net.cemetech.sfgp.scanline.geom.Edge.EndPoint;
import net.cemetech.sfgp.scanline.geom.Point.CoordName;

public class Primitive implements Comparable<Primitive> {
	public LinkedList<Edge> boundary;
	public Color color;
	public int unique;
	
	public static int uniqueV = 0;
	
	public Primitive(LinkedList<Edge> b){
		boundary = b;
		color = Color.black;
		unique = uniqueV++;
	}
	
	public static Primitive makeLine(Edge e){
		LinkedList<Edge> le = new LinkedList<Edge>();
		le.add(e);
		return new Primitive(le);
	}
	
	public static Primitive makeTri(Edge e1, Edge e2, Edge e3){
		LinkedList<Edge> le = new LinkedList<Edge>();
		le.add(e1);
		le.add(e2);
		le.add(e3);
		return new Primitive(le);
	}
	
	public static Primitive makeQuad(Edge e1, Edge e2, Edge e3, Edge e4){
		LinkedList<Edge> le = new LinkedList<Edge>();
		le.add(e1);
		le.add(e2);
		le.add(e3);
		le.add(e4);
		return new Primitive(le);
	}
	
	public int getArity(){ return boundary.size(); }
	
	public int getZForXY(int x, int y){
		assert(getArity() > 0);
		if(getArity() == 1){
			// Alpha blend! wut wut.
			Edge v = boundary.getFirst();
			Point vs = v.getEndPoint(EndPoint.START);
			Point ve = v.getEndPoint(EndPoint.END);
			
			int sx = vs.getComponent(CoordName.X);
			int sy = vs.getComponent(CoordName.Y);
			int sz = vs.getComponent(CoordName.Z);
			
			int dx = ve.getComponent(CoordName.X) - sx;
			int dy = ve.getComponent(CoordName.Y) - sy;
			int dz = ve.getComponent(CoordName.Z) - sz;
			
			assert((dx == 0 && dy == 0) || (dx != 0 && dy != 0));
			
			// Theoretically these should be the same, but
			// we'll average for robustness
			int xNumer = dz * (x - sx);
			int yNumer = dz * (y - sy);
			int xEst = (dx == 0) ? Math.min(0,  dz) : (xNumer / dx);
			int yEst = (dy == 0) ? Math.min(0,  dz) : (yNumer / dy);
			
			return sz + (xEst + yEst) / 2; 
		} else {
			Edge e1 = boundary.getFirst();
			Edge e2 = boundary.get(1);
			
			Point us = e1.getEndPoint(EndPoint.START);
			Point ue = e1.getEndPoint(EndPoint.END);
			Point u = new Point(
					us.getComponent(CoordName.X) - ue.getComponent(CoordName.X),
					us.getComponent(CoordName.Y) - ue.getComponent(CoordName.Y),
					us.getComponent(CoordName.Z) - ue.getComponent(CoordName.Z));
			int ux = u.getComponent(CoordName.X);
			int uy = u.getComponent(CoordName.Y);
			int uz = u.getComponent(CoordName.Z);
			
			Point vs = e2.getEndPoint(EndPoint.START);
			Point ve = e2.getEndPoint(EndPoint.END);
			Point v = new Point(
					vs.getComponent(CoordName.X) - ve.getComponent(CoordName.X),
					vs.getComponent(CoordName.Y) - ve.getComponent(CoordName.Y),
					vs.getComponent(CoordName.Z) - ve.getComponent(CoordName.Z));
			int vx = v.getComponent(CoordName.X);
			int vy = v.getComponent(CoordName.Y);
			int vz = v.getComponent(CoordName.Z);
			
			int nx = uy * vz - uz * vy;
			int ny = uz * vx - ux * vz;
			int nz = ux * vy - uy * vx;
			
			int d = -nx * us.getComponent(CoordName.X)
					-ny * us.getComponent(CoordName.Y)
					-nz * us.getComponent(CoordName.Z);
			
			int numer = (-d - nx * x - ny * y);
			
			return (nz == 0) ? ((numer > 0) ? Integer.MAX_VALUE : Integer.MIN_VALUE) : (numer / nz) ;
		}
	
	}
	//*
	public int hashCode() {
		int ret = 0;
		for(Edge e : boundary){
			ret ^= e.hashCode();
		}
		ret ^= color.hashCode();
		return ret;
	}
	//*/
	public int compareTo(Primitive o){
		return this.unique - o.unique;
	}

}
